File: packages/xdga_directories/src/xdga_directories.cpp
/*
 * Copyright (c) 2023. Open Mobile Platform LLC.
 * License: Proprietary.
 */
#include <QStandardPaths>

#include "xdga_directories.h"

char *getCacheLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::CacheLocation).toUtf8().data();
}

char *getAppDataLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::AppDataLocation).toUtf8().data();
}

char *getDocumentsLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation).toUtf8().data();
}

char *getDownloadLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::DownloadLocation).toUtf8().data();
}

char *getMusicLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::MusicLocation).toUtf8().data();
}

char *getPicturesLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::PicturesLocation).toUtf8().data();
}

char *getGenericDataLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation).toUtf8().data();
}

char *getMoviesLocation()
{
    return QStandardPaths::writableLocation(QStandardPaths::MoviesLocation).toUtf8().data();
}

File: packages/xdga_directories/lib/xdga_directories.dart
// SPDX-FileCopyrightText: 2023 Open Mobile Platform LLC <community@omp.ru>
// SPDX-License-Identifier: BSD-3-Clause
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import 'xdga_directories_bindings_generated.dart';

/// The dynamic library in which the symbols for [XdgaDirectoriesBindings] can be found.
final DynamicLibrary _dylib = () {
  return DynamicLibrary.open('libxdga_directories.so');
}();

/// The bindings to the native functions in [_dylib].
final XdgaDirectoriesBindings _bindings = XdgaDirectoriesBindings(_dylib);

/// QStandardPaths::CacheLocation
String getCacheLocation() =>
    _bindings.getCacheLocation().cast<Utf8>().toDartString();

/// QStandardPaths::AppDataLocation
String getAppDataLocation() =>
    _bindings.getAppDataLocation().cast<Utf8>().toDartString();

///  QStandardPaths::DocumentsLocation
String getDocumentsLocation() =>
    _bindings.getDocumentsLocation().cast<Utf8>().toDartString();

///  QStandardPaths::DownloadLocation
String getDownloadLocation() =>
    _bindings.getDownloadLocation().cast<Utf8>().toDartString();

///  QStandardPaths::MusicLocation
String getMusicLocation() =>
    _bindings.getMusicLocation().cast<Utf8>().toDartString();

///  QStandardPaths::PicturesLocation
String getPicturesLocation() =>
    _bindings.getPicturesLocation().cast<Utf8>().toDartString();

///  QStandardPaths::GenericDataLocation
String getGenericDataLocation() =>
    _bindings.getGenericDataLocation().cast<Utf8>().toDartString();

///  QStandardPaths::MoviesLocation
String getMoviesLocation() =>
    _bindings.getMoviesLocation().cast<Utf8>().toDartString();
